rm-datasource/src/main/java/io/seata/rm/datasource/AsyncWorker.java

```diff
@@ -171,12 +175,28 @@ private void doBranchCommits() {
                     continue;
                 }
                 List<Phase2Context> contextsGroupedByResourceId = entry.getValue();
+                Set<String> xids = new LinkedHashSet<>(UNDOLOG_DELETE_LIMIT_SIZE);
+                Set<Long> branchIds = new LinkedHashSet<>(UNDOLOG_DELETE_LIMIT_SIZE);
                 for (Phase2Context commitContext : contextsGroupedByResourceId) {
+                    xids.add(commitContext.xid);
+                    branchIds.add(commitContext.branchId);
+                    int maxSize = xids.size() > branchIds.size() ? xids.size() : branchIds.size();
+                    if(maxSize == UNDOLOG_DELETE_LIMIT_SIZE){
+                        try {
+                            UndoLogManager.batchDeleteUndoLog(xids, branchIds, UNDOLOG_DELETE_LIMIT_SIZE, conn);
+                        } catch (Exception ex) {
+                            LOGGER.warn("Failed to batch delete undo log [" +  "/" + xids + "]", ex);
+                        }
+                        xids.clear();
+                        branchIds.clear();
+                    }
+                }
+
+                if (CollectionUtils.isNotEmpty(xids) && CollectionUtils.isNotEmpty(branchIds)) {
```

Comment:
if (CollectionUtils.isEmpty(xids) || CollectionUtils.isEmpty(branchIds)){
continue;
}
reduce cyclomatic complexity

Reply:
i can ‘t understand ？ the code is out of for{}

Reply:
done